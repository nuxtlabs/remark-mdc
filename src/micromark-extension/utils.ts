import type { Effects } from 'micromark-factory-space'

// Measure the number of character codes in chunks.
// Counts tabs based on their expanded size, and CR+LF as one character.
export function sizeChunks (chunks: any[]) {
  let index = -1
  let size = 0

  while (++index < chunks.length) {
    size += typeof chunks[index] === 'string' ? chunks[index].length : 1
  }

  return size
}

export function prefixSize (events: any, type: string) {
  const tail = events[events.length - 1]
  if (!tail || tail[1].type !== type) { return 0 }
  return sizeChunks(tail[2].sliceStream(tail[1]))
}

/**
 * Calculate line indention size, line indention could be consists of multiple `linePrefix` events
 * @param events parser tokens
 * @returns line indention size
 */
export function linePrefixSize (events: any[]) {
  let size = 0
  let index = events.length - 1
  let tail = events[index]
  while (index >= 0 && tail && tail[1].type === 'linePrefix' && tail[0] === 'exit') {
    size += sizeChunks(tail[2].sliceStream(tail[1]))
    index -= 1
    tail = events[index]
  }

  return size
}

/**
 * Manage token state
 */
export const useTokenState = (tokenName: string) => {
  const token = {
    isOpen: false,
    /**
     * Enter into token, close previous open token if any
     */
    enter: (effects: Effects) => {
      const initialState = token.isOpen
      token.exit(effects)
      effects.enter(tokenName)
      token.isOpen = true

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Enter into token only once, if token is already open, do nothing
     */
    enterOnce: (effects: Effects) => {
      const initialState = token.isOpen

      if (!token.isOpen) {
        effects.enter(tokenName)
        token.isOpen = true
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Exit from token if it is open
     */
    exit: (effects: Effects) => {
      const initialState = token.isOpen
      if (token.isOpen) {
        effects.exit(tokenName)
        token.isOpen = false
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    }
  }
  return token
}
