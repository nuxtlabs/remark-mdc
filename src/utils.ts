export const NON_UNWRAPABLE_TYPES = [
  'componentContainerSection',
  'componentContainerDataSection',
  'containerComponent',
  'leafComponent'
]
