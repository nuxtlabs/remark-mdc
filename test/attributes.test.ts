import { describe } from 'vitest'
import { runMarkdownTests } from './utils'

describe('Attributes', () => {
  runMarkdownTests({
    'id-suger': {
      markdown: ':test{#id} text',
    },
    'class-suger': {
      markdown: ':test{.class .another-class key="value"} text',
      expected: ':test{.class.another-class key="value"} text',
    },
    'boolean': {
      markdown: ':test{prop} text',
      expected: ':test{prop="true"} text',
    },
    'html-characters': {
      markdown: ':test{icon="&copy;"} text',
      expected: ':test{icon="©"} text',
    },
    'html-characters-unquoted': {
      markdown: ':test{icon=&copy;} text',
      expected: ':test{icon="©"} text',
    },
    'value-with-space': {
      markdown: ':test{attr= value} text',
      expected: ':test{attr="value"} text',
    },
    'invalid-binding': {
      markdown: ':test{:} text',
      expected: ':test{:} text',
    },
    'fragment-attribute': {
      markdown: '[![Nuxt](https://nuxtjs.org/design-kit/colored-logo.svg){.nest}](https:test){.cls}',
      expected: '[![Nuxt](https://nuxtjs.org/design-kit/colored-logo.svg){.nest}](https\\:test){.cls}',
    },
    'image': {
      markdown: '![Nuxt](https://nuxtjs.org/design-kit/colored-logo.svg){#id .class}',
    },
    'nested': {
      markdown: '[Hello [World]{.x}](/world){style="color: green"}',
    },
    'nested2': {
      markdown: '[Hello [World](/world){.x}]{style="color: green"}',
    },
    'nested3': {
      markdown: '_**[Nuxt](https://nuxtjs.org){#id .class} strong**{#id2 .class2} emphasis_{#id3 .class3}',
    },
    'code': {
      markdown: '`code`{#id .class}',
    },
    'strong': {
      markdown: '**strong**{#id .class}',
    },
    'link': {
      markdown: '[Nuxt](https://nuxtjs.org){#id .class}',
    },
    'linkReference': {
      markdown: '[Nuxt][nuxt]{#id .class}\n\n[nuxt]: https://nuxt.com',
    },
    'emphasis': {
      markdown: '_emphasis_{#id .class}',
    },
    'nested-in-table': {
      markdown: [
        '| Col1 |      Col2      |',
        '|  --  |      -----     |',
        '|  aa  | [a](/a){a="a"} |',
      ].join('\n'),
      expected: [
        '| Col1 | Col2           |',
        '| ---- | -------------- |',
        '| aa   | [a](/a){a="a"} |',
      ].join('\n'),
    },
  })
})
