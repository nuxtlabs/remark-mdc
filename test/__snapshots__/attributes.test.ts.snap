// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Attributes > boolean 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "prop": true,
          },
          "children": [],
          "data": {
            "hName": "test",
            "hProperties": {
              ":prop": "true",
            },
          },
          "fmAttributes": {},
          "name": "test",
          "position": {
            "end": {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
        {
          "position": {
            "end": {
              "column": 17,
              "line": 1,
              "offset": 16,
            },
            "start": {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
          },
          "type": "text",
          "value": " text",
        },
      ],
      "position": {
        "end": {
          "column": 17,
          "line": 1,
          "offset": 16,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 17,
      "line": 1,
      "offset": 16,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > class-suger 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "class": "class another-class",
            "key": "value",
          },
          "children": [],
          "data": {
            "hName": "test",
            "hProperties": {
              "class": "class another-class",
              "key": "value",
            },
          },
          "fmAttributes": {},
          "name": "test",
          "position": {
            "end": {
              "column": 41,
              "line": 1,
              "offset": 40,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
        {
          "position": {
            "end": {
              "column": 46,
              "line": 1,
              "offset": 45,
            },
            "start": {
              "column": 41,
              "line": 1,
              "offset": 40,
            },
          },
          "type": "text",
          "value": " text",
        },
      ],
      "position": {
        "end": {
          "column": 46,
          "line": 1,
          "offset": 45,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 46,
      "line": 1,
      "offset": 45,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > code 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "class": "class",
            "id": "id",
          },
          "position": {
            "end": {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "inlineCode",
          "value": "code",
        },
      ],
      "position": {
        "end": {
          "column": 19,
          "line": 1,
          "offset": 18,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 19,
      "line": 1,
      "offset": 18,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > emphasis 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "class": "class",
            "id": "id",
          },
          "children": [
            {
              "position": {
                "end": {
                  "column": 10,
                  "line": 1,
                  "offset": 9,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "emphasis",
            },
          ],
          "position": {
            "end": {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "emphasis",
        },
      ],
      "position": {
        "end": {
          "column": 23,
          "line": 1,
          "offset": 22,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 23,
      "line": 1,
      "offset": 22,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > fragment-attribute 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "class": "cls",
          },
          "children": [
            {
              "alt": "Nuxt",
              "attributes": {
                "class": "nest",
              },
              "position": {
                "end": {
                  "column": 57,
                  "line": 1,
                  "offset": 56,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "title": null,
              "type": "image",
              "url": "https://nuxtjs.org/design-kit/colored-logo.svg",
            },
          ],
          "position": {
            "end": {
              "column": 77,
              "line": 1,
              "offset": 76,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "title": null,
          "type": "link",
          "url": "https:test",
        },
      ],
      "position": {
        "end": {
          "column": 83,
          "line": 1,
          "offset": 82,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 83,
      "line": 1,
      "offset": 82,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > html-characters 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "icon": "©",
          },
          "children": [],
          "data": {
            "hName": "test",
            "hProperties": {
              "icon": "©",
            },
          },
          "fmAttributes": {},
          "name": "test",
          "position": {
            "end": {
              "column": 21,
              "line": 1,
              "offset": 20,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
        {
          "position": {
            "end": {
              "column": 26,
              "line": 1,
              "offset": 25,
            },
            "start": {
              "column": 21,
              "line": 1,
              "offset": 20,
            },
          },
          "type": "text",
          "value": " text",
        },
      ],
      "position": {
        "end": {
          "column": 26,
          "line": 1,
          "offset": 25,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 26,
      "line": 1,
      "offset": 25,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > html-characters-unquoted 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "icon": "©",
          },
          "children": [],
          "data": {
            "hName": "test",
            "hProperties": {
              "icon": "©",
            },
          },
          "fmAttributes": {},
          "name": "test",
          "position": {
            "end": {
              "column": 19,
              "line": 1,
              "offset": 18,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
        {
          "position": {
            "end": {
              "column": 24,
              "line": 1,
              "offset": 23,
            },
            "start": {
              "column": 19,
              "line": 1,
              "offset": 18,
            },
          },
          "type": "text",
          "value": " text",
        },
      ],
      "position": {
        "end": {
          "column": 24,
          "line": 1,
          "offset": 23,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 24,
      "line": 1,
      "offset": 23,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > id-suger 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "id": "id",
          },
          "children": [],
          "data": {
            "hName": "test",
            "hProperties": {
              "id": "id",
            },
          },
          "fmAttributes": {},
          "name": "test",
          "position": {
            "end": {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
        {
          "position": {
            "end": {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
            "start": {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
          },
          "type": "text",
          "value": " text",
        },
      ],
      "position": {
        "end": {
          "column": 16,
          "line": 1,
          "offset": 15,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 16,
      "line": 1,
      "offset": 15,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > ignoreEscapeCharacterInNormalAttribute 1`] = `
{
  "children": [
    {
      "attributes": {
        "code": "D:\\Software\\",
      },
      "children": [],
      "data": {
        "hName": "copy",
        "hProperties": {
          "code": "D:\\Software\\",
        },
      },
      "fmAttributes": {},
      "name": "copy",
      "position": {
        "end": {
          "column": 27,
          "line": 1,
          "offset": 26,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "textComponent",
    },
  ],
  "position": {
    "end": {
      "column": 27,
      "line": 1,
      "offset": 26,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > ignoreEscapeCharacterInNormalAttributeYaml 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [],
      "data": {
        "hName": "copy",
        "hProperties": {
          "code": "D:\\Software\\",
        },
      },
      "fmAttributes": {
        "code": "D:\\Software\\",
      },
      "name": "copy",
      "position": {
        "end": {
          "column": 3,
          "line": 5,
          "offset": 36,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": "
code: D:\\Software\\
---",
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 5,
      "offset": 36,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > image 1`] = `
{
  "children": [
    {
      "children": [
        {
          "alt": "Nuxt",
          "attributes": {
            "class": "class",
            "id": "id",
          },
          "position": {
            "end": {
              "column": 56,
              "line": 1,
              "offset": 55,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "title": null,
          "type": "image",
          "url": "https://nuxtjs.org/design-kit/colored-logo.svg",
        },
      ],
      "position": {
        "end": {
          "column": 68,
          "line": 1,
          "offset": 67,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 68,
      "line": 1,
      "offset": 67,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > invalid-binding 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {},
          "children": [],
          "data": {
            "hName": "test",
            "hProperties": {},
          },
          "fmAttributes": {},
          "name": "test",
          "position": {
            "end": {
              "column": 6,
              "line": 1,
              "offset": 5,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
        {
          "position": {
            "end": {
              "column": 14,
              "line": 1,
              "offset": 13,
            },
            "start": {
              "column": 6,
              "line": 1,
              "offset": 5,
            },
          },
          "type": "text",
          "value": "{:} text",
        },
      ],
      "position": {
        "end": {
          "column": 14,
          "line": 1,
          "offset": 13,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 14,
      "line": 1,
      "offset": 13,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > link 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "class": "class",
            "id": "id",
          },
          "children": [
            {
              "position": {
                "end": {
                  "column": 6,
                  "line": 1,
                  "offset": 5,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "Nuxt",
            },
          ],
          "position": {
            "end": {
              "column": 27,
              "line": 1,
              "offset": 26,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "title": null,
          "type": "link",
          "url": "https://nuxtjs.org",
        },
      ],
      "position": {
        "end": {
          "column": 39,
          "line": 1,
          "offset": 38,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 39,
      "line": 1,
      "offset": 38,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > linkReference 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "class": "class",
            "id": "id",
          },
          "children": [
            {
              "position": {
                "end": {
                  "column": 6,
                  "line": 1,
                  "offset": 5,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "Nuxt",
            },
          ],
          "identifier": "nuxt",
          "label": "nuxt",
          "position": {
            "end": {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "referenceType": "full",
          "type": "linkReference",
        },
      ],
      "position": {
        "end": {
          "column": 25,
          "line": 1,
          "offset": 24,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
    {
      "identifier": "nuxt",
      "label": "nuxt",
      "position": {
        "end": {
          "column": 25,
          "line": 3,
          "offset": 50,
        },
        "start": {
          "column": 1,
          "line": 3,
          "offset": 26,
        },
      },
      "title": null,
      "type": "definition",
      "url": "https://nuxt.com",
    },
  ],
  "position": {
    "end": {
      "column": 25,
      "line": 3,
      "offset": 50,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > nested 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "style": "color: green",
          },
          "children": [
            {
              "position": {
                "end": {
                  "column": 8,
                  "line": 1,
                  "offset": 7,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "Hello ",
            },
            {
              "attributes": {
                "class": "x",
              },
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 14,
                      "line": 1,
                      "offset": 13,
                    },
                    "start": {
                      "column": 9,
                      "line": 1,
                      "offset": 8,
                    },
                  },
                  "type": "text",
                  "value": "World",
                },
              ],
              "data": {
                "hName": "span",
                "hProperties": {
                  "class": "x",
                },
              },
              "fmAttributes": {},
              "name": "span",
              "position": {
                "end": {
                  "column": 15,
                  "line": 1,
                  "offset": 14,
                },
                "start": {
                  "column": 8,
                  "line": 1,
                  "offset": 7,
                },
              },
              "type": "textComponent",
            },
          ],
          "position": {
            "end": {
              "column": 28,
              "line": 1,
              "offset": 27,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "title": null,
          "type": "link",
          "url": "/world",
        },
      ],
      "position": {
        "end": {
          "column": 50,
          "line": 1,
          "offset": 49,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 50,
      "line": 1,
      "offset": 49,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > nested-in-table 1`] = `
{
  "children": [
    {
      "align": [
        null,
        null,
      ],
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 7,
                      "line": 1,
                      "offset": 6,
                    },
                    "start": {
                      "column": 3,
                      "line": 1,
                      "offset": 2,
                    },
                  },
                  "type": "text",
                  "value": "Col1",
                },
              ],
              "position": {
                "end": {
                  "column": 8,
                  "line": 1,
                  "offset": 7,
                },
                "start": {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "tableCell",
            },
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 19,
                      "line": 1,
                      "offset": 18,
                    },
                    "start": {
                      "column": 15,
                      "line": 1,
                      "offset": 14,
                    },
                  },
                  "type": "text",
                  "value": "Col2",
                },
              ],
              "position": {
                "end": {
                  "column": 26,
                  "line": 1,
                  "offset": 25,
                },
                "start": {
                  "column": 8,
                  "line": 1,
                  "offset": 7,
                },
              },
              "type": "tableCell",
            },
          ],
          "position": {
            "end": {
              "column": 26,
              "line": 1,
              "offset": 25,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "tableRow",
        },
        {
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 6,
                      "line": 3,
                      "offset": 57,
                    },
                    "start": {
                      "column": 4,
                      "line": 3,
                      "offset": 55,
                    },
                  },
                  "type": "text",
                  "value": "aa",
                },
              ],
              "position": {
                "end": {
                  "column": 8,
                  "line": 3,
                  "offset": 59,
                },
                "start": {
                  "column": 1,
                  "line": 3,
                  "offset": 52,
                },
              },
              "type": "tableCell",
            },
            {
              "children": [
                {
                  "attributes": {
                    "a": "a",
                  },
                  "children": [
                    {
                      "position": {
                        "end": {
                          "column": 12,
                          "line": 3,
                          "offset": 63,
                        },
                        "start": {
                          "column": 11,
                          "line": 3,
                          "offset": 62,
                        },
                      },
                      "type": "text",
                      "value": "a",
                    },
                  ],
                  "position": {
                    "end": {
                      "column": 17,
                      "line": 3,
                      "offset": 68,
                    },
                    "start": {
                      "column": 10,
                      "line": 3,
                      "offset": 61,
                    },
                  },
                  "title": null,
                  "type": "link",
                  "url": "/a",
                },
              ],
              "position": {
                "end": {
                  "column": 26,
                  "line": 3,
                  "offset": 77,
                },
                "start": {
                  "column": 8,
                  "line": 3,
                  "offset": 59,
                },
              },
              "type": "tableCell",
            },
          ],
          "position": {
            "end": {
              "column": 26,
              "line": 3,
              "offset": 77,
            },
            "start": {
              "column": 1,
              "line": 3,
              "offset": 52,
            },
          },
          "type": "tableRow",
        },
      ],
      "position": {
        "end": {
          "column": 26,
          "line": 3,
          "offset": 77,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "table",
    },
  ],
  "position": {
    "end": {
      "column": 26,
      "line": 3,
      "offset": 77,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > nested2 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "style": "color: green",
          },
          "children": [
            {
              "position": {
                "end": {
                  "column": 8,
                  "line": 1,
                  "offset": 7,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "Hello ",
            },
            {
              "attributes": {
                "class": "x",
              },
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 14,
                      "line": 1,
                      "offset": 13,
                    },
                    "start": {
                      "column": 9,
                      "line": 1,
                      "offset": 8,
                    },
                  },
                  "type": "text",
                  "value": "World",
                },
              ],
              "position": {
                "end": {
                  "column": 23,
                  "line": 1,
                  "offset": 22,
                },
                "start": {
                  "column": 8,
                  "line": 1,
                  "offset": 7,
                },
              },
              "title": null,
              "type": "link",
              "url": "/world",
            },
          ],
          "data": {
            "hName": "span",
            "hProperties": {
              "style": "color: green",
            },
          },
          "fmAttributes": {},
          "name": "span",
          "position": {
            "end": {
              "column": 28,
              "line": 1,
              "offset": 27,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
      ],
      "position": {
        "end": {
          "column": 50,
          "line": 1,
          "offset": 49,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 50,
      "line": 1,
      "offset": 49,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > nested3 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "class": "class3",
            "id": "id3",
          },
          "children": [
            {
              "attributes": {
                "class": "class2",
                "id": "id2",
              },
              "children": [
                {
                  "attributes": {
                    "class": "class",
                    "id": "id",
                  },
                  "children": [
                    {
                      "position": {
                        "end": {
                          "column": 9,
                          "line": 1,
                          "offset": 8,
                        },
                        "start": {
                          "column": 5,
                          "line": 1,
                          "offset": 4,
                        },
                      },
                      "type": "text",
                      "value": "Nuxt",
                    },
                  ],
                  "position": {
                    "end": {
                      "column": 30,
                      "line": 1,
                      "offset": 29,
                    },
                    "start": {
                      "column": 4,
                      "line": 1,
                      "offset": 3,
                    },
                  },
                  "title": null,
                  "type": "link",
                  "url": "https://nuxtjs.org",
                },
                {
                  "position": {
                    "end": {
                      "column": 49,
                      "line": 1,
                      "offset": 48,
                    },
                    "start": {
                      "column": 42,
                      "line": 1,
                      "offset": 41,
                    },
                  },
                  "type": "text",
                  "value": " strong",
                },
              ],
              "position": {
                "end": {
                  "column": 51,
                  "line": 1,
                  "offset": 50,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "strong",
            },
            {
              "position": {
                "end": {
                  "column": 74,
                  "line": 1,
                  "offset": 73,
                },
                "start": {
                  "column": 65,
                  "line": 1,
                  "offset": 64,
                },
              },
              "type": "text",
              "value": " emphasis",
            },
          ],
          "position": {
            "end": {
              "column": 75,
              "line": 1,
              "offset": 74,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "emphasis",
        },
      ],
      "position": {
        "end": {
          "column": 89,
          "line": 1,
          "offset": 88,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 89,
      "line": 1,
      "offset": 88,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > strong 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "class": "class",
            "id": "id",
          },
          "children": [
            {
              "position": {
                "end": {
                  "column": 9,
                  "line": 1,
                  "offset": 8,
                },
                "start": {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "type": "text",
              "value": "strong",
            },
          ],
          "position": {
            "end": {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "strong",
        },
      ],
      "position": {
        "end": {
          "column": 23,
          "line": 1,
          "offset": 22,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 23,
      "line": 1,
      "offset": 22,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Attributes > value-with-space 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "attr": "value",
          },
          "children": [],
          "data": {
            "hName": "test",
            "hProperties": {
              "attr": "value",
            },
          },
          "fmAttributes": {},
          "name": "test",
          "position": {
            "end": {
              "column": 19,
              "line": 1,
              "offset": 18,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "textComponent",
        },
        {
          "position": {
            "end": {
              "column": 24,
              "line": 1,
              "offset": 23,
            },
            "start": {
              "column": 19,
              "line": 1,
              "offset": 18,
            },
          },
          "type": "text",
          "value": " text",
        },
      ],
      "position": {
        "end": {
          "column": 24,
          "line": 1,
          "offset": 23,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 24,
      "line": 1,
      "offset": 23,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
